### ISSUE TITLE
Colors in the product data network

### ISSUE THREAD

**Issue by christoph-maurer** [building-envelope-data/api#323]
Currently, the product data network uses the [CIELAB color space](https://en.wikipedia.org/wiki/CIELAB_color_space) ([example](https://github.com/building-envelope-data/api/blob/3f4d20eeb5a2e45fcf4ba0c51f90c7722965f228/tests/valid/opticalData/SunovationModuleEnamelOptical.json#L4), [schema](https://github.com/building-envelope-data/api/blob/3f4d20eeb5a2e45fcf4ba0c51f90c7722965f228/schemas/opticalData.json#L392)) and IGSDB uses the [CIEXYZ color space]. There are calculations available to convert colors from CIELAB to CIEXYZ and reverse.

@RDmitchell @danielmcquillen If Fraunhofer changes from CIELAB to CIEXYZ, would it be easy for you to offer X, Y and Z in the same way as you already offer [nearnormalHemisphericalVisibleTransmittances](https://github.com/building-envelope-data/api/blob/3f4d20eeb5a2e45fcf4ba0c51f90c7722965f228/queries/database/tutorial.graphql#L75)?

**Comment by danielmcquillen** [comment]
@christoph-maurer In IGSDBv2, every product provides color information in three formats:

For example, the coated product at  https://igsdb-v2.herokuapp.com/products/low-e-6/

<img width="882" alt="image" src="https://github.com/building-envelope-data/api/assets/119496/ccd331cf-e09e-4f52-9f61-5e45fcb7b6fb">

**Comment by danielmcquillen** [comment]
The json structure provided by the api provides a scaffolding for more information about color. However, at the moment most products (which have been directly imported from the IGDB) only have a few fields defined.

This information is located in the dictionaries within the `integrated_spectral_averages_summaries` list (there's usually only one dictionary, representing the summary values for the `NFRC` standard.

This dictionary includes the `color` key, which contains `lab`, 'rgb', and `trichromatic` data for each combination of reflectance front/back and direct/diffuse combinations:

<img width="739" alt="image" src="https://github.com/building-envelope-data/api/assets/119496/16424020-110a-461a-bd5b-e97e671688f6">


For example:

https://igsdb-v2.herokuapp.com/api/v2/products/clr-3/

**Comment by christoph-maurer** [comment]
@danielmcquillen I like how you attach color to the integrated_spectral_averages! 

When I move ["color"](https://github.com/building-envelope-data/api/blob/3f4d20eeb5a2e45fcf4ba0c51f90c7722965f228/schemas/opticalData.json#L363-L364) to the ["results"](https://github.com/building-envelope-data/api/blob/3f4d20eeb5a2e45fcf4ba0c51f90c7722965f228/tests/valid/opticalData/igsdbExample196.json#L2998-L2999), would it be easy for you to include it into BED-JSON data sets? Would it also be easy to return the CIE L\*a\*b\* values in [GraphQL](https://github.com/building-envelope-data/api/blob/3f4d20eeb5a2e45fcf4ba0c51f90c7722965f228/apis/database.graphql#L1188-L1189)?

**Comment by danielmcquillen** [comment]
@christoph-maurer Sure, I can look into that this week. It's probably easier to update the BED-JSON output so I'll look at that first.

**Comment by danielmcquillen** [comment]
@christoph-maurer Can you provide an example product json in `BED-JSON` format that demostrates where the color vales from the integrated summary values would go?

**Comment by christoph-maurer** [comment]
@danielmcquillen Yes, I'll work on it. I will send it to you as soon as it is ready.

**Comment by danielmcquillen** [comment]
@christoph-maurer Awesome. I'll work with @RDmitchell and Jacob to update once you've got some guidance ready! (By the way, would it make sense to add Jacob to your repo so he can comment? https://github.com/jcjonsson ... Robin is out of the office for a couple weeks.)

**Comment by christoph-maurer** [comment]
@danielmcquillen Excellent idea! I have added @jcjonsson and @RDmitchell to the team [EQWIN-
P_associated_partners](https://github.com/orgs/building-envelope-data/teams/eqwin-p_associated_partners/members). This should give them access to all relevant repositories.

**Comment by christoph-maurer** [comment]
@danielmcquillen I have moved the keys for color in the schema opticalData.json to ["results"](https://github.com/building-envelope-data/api/blob/a9ce6beca5dae053da8e23f7c072c0204cf600b0/schemas/opticalData.json#L74-L100). I have created a [test](https://github.com/building-envelope-data/api/blob/a9ce6beca5dae053da8e23f7c072c0204cf600b0/tests/valid/opticalData/colorIgsdbExampleClr-3.json#L23-L39) as example for color data based on https://igsdb-v2.herokuapp.com/api/v2/products/clr-3/ . The [API specification for databases](https://github.com/building-envelope-data/api/blob/a9ce6beca5dae053da8e23f7c072c0204cf600b0/apis/database.graphql#L1167-L1185) was already including fields for color.

It is nice when the product data network can be queried for color. But it has a low priority, because it is not currently needed by a project partner.

**Comment by danielmcquillen** [comment]
@christoph-maurer Thanks for making the test. I'm confused though as the test seems to indicate color information available at every wavelength, yet in the IGSDB color is only available at a higher level: once for each standard used to create integrated summary values (at the moment, we only do this for the "NFRC" standard). 

We don't have color information at every wavelength. 

Here's a snippet of `clr-3` through the IGSDB v2 API in `bed-json` format showing the measured data at 300nm:

<img width="632" alt="image" src="https://github.com/user-attachments/assets/3fd3eea7-10e0-43cd-8767-064f72b2dec9">

I would have thought any color information would appear as a property of `componentCharacteristics`, not futher down into the actual `dataPoints` list.

Furthermore, I noticed in your test example you have this:

<img width="481" alt="image" src="https://github.com/user-attachments/assets/5f92ad3a-f97f-4b9d-b715-087676127696">

Yet the IGSDB implementation has a `wavelength` field with the actual wavelength in nm as a value. Is there something amiss here or is this all ok?

**Comment by christoph-maurer** [comment]
Hi @danielmcquillen , https://igsdb-v2.herokuapp.com/api/v2/products/clr-3/?format=bed-json currently contains `dataPoints` for single wavelengths like
```
        {
          "incidence": {
            "wavelengths": {
              "wavelength": 300
            },
            "direction": {
              "polar": 8
            }
          },
          "emergence": {
            "direction": {
              "polar": 8
            }
          },
          "results": {
            "transmittance": "0.0",
            "reflectance": "0.0467"
          }
        },
```
as well as `dataPoints` integrated over a spectrum like
```
        {
          "incidence": {
            "wavelengths": {
              "integral": "visible"
            },
            "direction": {
              "polar": 8
            }
          },
          "emergence": {
            "direction": "hemispherical"
          },
          "results": {
            "transmittance": 0.9044294,
            "reflectance": 0.08225197
          }
        },
```
`wavelengths` can either have the key `wavelength` to define a single wavelength or the key `integral` to define a spectrum over which the results are integrated.

My proposal is that you extend integrals over the visible spectrum with the color data. For example, you can change the last example to:
```
        {
          "incidence": {
            "wavelengths": {
              "integral": "visible"
            },
            "direction": {
              "polar": 8
            }
          },
          "emergence": {
            "direction": "hemispherical"
          },
          "results": {
            "transmittance": 0.9044294,
            "colorTransmittance": {
              "cielabColors": [
                {
                  "aStar": -0.8035122491914515,
                  "bStar": 0.2168086516082468,
                  "lStar": 96.18335944095739
                }
              ]
            },
            "reflectance": 0.08225287,
            "colorReflectance": {
              "cielabColors": [
                {
                  "aStar": -0.4781490764729779,
                  "bStar": -0.4878317966566348,
                  "lStar": 34.46499873850816
                }
              ]
            }
          }
        }
```

The schema `opticalData.json` allows the keys `colorTransmittance` and `colorReflectance` [only if](https://github.com/building-envelope-data/api/blob/c9c2c3d9098d3f9a2a0698cc7c76bf2c30b41d91/schemas/opticalData.json#L74-L90)
```
            "wavelengths": {
              "integral": "visible"
            },
```
It is therefore not possible to define the color for a single wavelength or for another spectrum than `visible`.

@danielmcquillen Does this help to clarify the issue?

**Comment by danielmcquillen** [comment]
Thanks for the clarification, Christophe. Very helpful. Traveling this week but will update the serializer as soon as I return next week.

**Comment by danielmcquillen** [comment]
@christoph-maurer @simon-wacker We would need to discuss this update with @jcjonsson in order to understand what would be involved to implement this.

**Comment by christoph-maurer** [comment]
@danielmcquillen Do you discuss the update with @jcjonsson or would you like to propose a date for a web conference with @jcjonsson @danielmcquillen and me?

**Comment by simon-wacker** [comment]
@christoph-maurer @danielmcquillen @jcjonsson If there are CIELAB colors in the IGSDB, then they are filterable and are returned. This was implemented in https://github.com/LBNL-ETA/igsdb-webapp/pull/212 However, when I query the IGSDB for data with at least one CIELAB color entry, the result list is empty:

```
{
  allOpticalData(where: {cielabColors: {some: {lStar: {greaterThanOrEqualTo: 0}}}}) {
    edges {
      node {
        componentId
        cielabColors {
          lStar
          aStar
          bStar
        }
        resourceTree {
          root {
            value {
              locator
            }
          }
        }
      }
    }
  }
}
```

So, to serve the color values that exist in other formats like CIEXYZ, some work still needs to be done. One option would be to save colors in as many color spaces as possible in the IGSDB (for example by converting them automatically when a data set is added). The other option, to convert color values on the fly when accessed via GraphQL is in my opinion more tedious and slows down the system because it would require conversion in both directions for every request and every data set (in one direction when filtering and in the other when delivering data).

**Comment by danielmcquillen** [comment]
@christoph-maurer @simon-wacker I discussed this issue with @jcjonsson and @RDmitchell today.

To recap:

IGSDBv2 has spots for XYZ, RGB and LAB color definitions in the `integrated_spectral_averages_summary` dictionary in a product definition for transmittance and reflectance front and back. The team doesn't intend on supporting other colour formats.

(See the API for example: https://igsdb-v2.herokuapp.com/api/v2/products/clr-3)

So if there are further colour definitions you want to filter on, you'd have to generate those runtime before filtering, per @simon-wacker 's comment.

Note also that the current data is only a subset of what we later expect to have: because current data comes directly from the IGDB we can only fill in a subset of the color values. When Checkertool generates this information via Pywincalc directly, and this information is published to the IGSDB directly from Checkertool, more of these properties will be populated where they currently show `null`....but again, only those three formats: XYZ, RGB and LAB.

@simon-wacker If you want to enable searching on those three existing formats and the limited subset of actual values available for all products, I think we'd have to push them into the materialized table.

**Comment by christoph-maurer** [comment]
@simon-wacker In the past, we thought that the optical data should not be redundant in a ways that one value can easily be derived from another value. I think that it's sufficient that the CIELAB color space is the only one which can be used for queries.

@jcjonsson Do you agree that the CIELAB color coordinates can easily be converted from and into other color spaces?

**Comment by danielmcquillen** [comment]
@simon-wacker Yes understood about LAB. My impression was there was somehow other color formats what were somehow possible and you wanted to support. If it's only XYZ, RGB and LAB I think we're good.

More robust color values won't show up until we start using Checkertool to submit products to the IGSDB, as Checkertool uses Pywincalc which produces more data than what you see in the IGDB-derived data.

**Comment by danielmcquillen** [comment]
Hmm. I would defer to @RDmitchell @jcjonsson for an estimate.

**Comment by RDmitchell** [comment]
@christoph-maurer -- it's a bit hard to say. Hopefully in the next 6 months.
